code/go/0chain.net/smartcontract/storagesc/shut_down.go

```diff
@@ -0,0 +1,83 @@
+package storagesc
+
+import (
+	cstate "0chain.net/chaincore/chain/state"
+	"0chain.net/chaincore/transaction"
+	"0chain.net/core/common"
+)
+
+func (ssc *StorageSmartContract) shutDownBlobber(
+	t *transaction.Transaction,
+	input []byte,
+	balances cstate.StateContextI,
+) (string, error) {
+	var req providerRequest
+	if err := req.decode(input); err != nil {
+		return "", common.NewError("shut_down_blobber_failed", err.Error())
+	}
+
+	var err error
+	var blobber *StorageNode
+	if blobber, err = ssc.getBlobber(req.ID, balances); err != nil {
+		return "", common.NewError("shut_down_blobber_failed",
+			"can't get the blobber "+t.ClientID+": "+err.Error())
+	}
+	if t.ClientID != blobber.StakePoolSettings.DelegateWallet {
+		return "", common.NewError("shut_down_blobber_failed",
+			"access denied, allowed for delegate_wallet owner only")
+	}
+
+	blobber.ShutDown()
+	if err := emitUpdateBlobber(blobber, balances); err != nil {
+		return "", common.NewError("shut_down_blobber_failed", err.Error())
+	}
+	if _, err := balances.InsertTrieNode(blobber.GetKey(ssc.ID), blobber); err != nil {
+		return "", common.NewError("shut_down_blobber_failed",
+			"can't save blobber: "+err.Error())
+	}
+	return "", nil
+}
+
+func (ssc *StorageSmartContract) shutDownValidator(
+	t *transaction.Transaction,
+	input []byte,
+	balances cstate.StateContextI,
+) (string, error) {
```

Comment:
generalize common code

Reply:
The only common code is decoding the provided request. We could use more common code, but that would mean adding get and save provider methods. Which would mean using a common format for the MPT key. Maybe a good idea but should we do that here?